kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+                    });
+                }
+                break ;
+            } else {
+                proof {
+                    assert(match cur {
+                        Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                            (self.index@ - 1) as nat,
+                        ) && cur == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                        Option::None => false,
+                    });
+                }
+                last = cur;
+            }
+        }
+        last
+    }
+}
+
+impl<'a, T: ?Sized + ListNodeV<'a, T>> ListV<'a, T> {
+    pub closed spec fn well_formed_list(&self, ghost_state: &Tracked<GhostState<'a, T>>) -> bool {
+        &&& ghost_state@.cells.len() >= 1
+        &&& forall|i: nat| 0 <= i < ghost_state@.cells.len() ==> well_formed_node(ghost_state, i)
+        &&& forall|i: nat|
+            0 <= i < (ghost_state@.cells.len() - 1) as nat
+                ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                Option::Some(_) => true,
+                Option::None => false,
+            }
+        &&& match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+            Option::Some(_) => false,
+            Option::None => true,
+        }
+        &&& ghost_state@.cells[0].id() == self.head.0.id()
+    }
+
+    pub const fn new() -> (res: (ListV<'a, T>, Tracked<GhostState<'a, T>>))
+        ensures
+            res.1@.cells.len() == 1,
+            well_formed_node(&res.1, 0),
+            res.0.well_formed_list(&res.1),
+    {
+        let (list_link, Tracked(points_to)) = ListLinkV::<T>::empty();
+        let cell = &list_link.0;
+        let mut ghost_state = Tracked(
+            GhostState { cells: Seq::empty(), points_to_map: Map::tracked_empty() },
+        );
+
+        proof {
+            ghost_state.borrow_mut().cells = ghost_state@.cells.push(cell);
+            ghost_state.borrow_mut().points_to_map.tracked_insert(0, points_to);
+        }
+
+        let list = ListV { head: list_link };
+
+        (list, ghost_state)
+    }
+
+    pub fn head(&self, ghost_state: &Tracked<GhostState<'a, T>>) -> (res: Option<&'a T>)
+        requires
+            ghost_state@.cells.len() >= 1,
+            ghost_state@.points_to_map.dom().contains(0),
+            ghost_state@.cells[0].id() == self.head.0.id(),
+            ghost_state@.points_to_map[0].is_init(),
+            ghost_state@.points_to_map[0].id() == self.head.0.id(),
+        ensures
+            ghost_state@.points_to_map[0].value() == res,
+    {
+        let tracked points_to = ghost_state.borrow().points_to_map.tracked_borrow(0);
+        *self.head.0.borrow(Tracked(points_to))
+    }
+
+    pub fn push_head(
+        &self,
+        node: &'a T,
+        next_points_to: Tracked<PointsTo<Option<&'a T>>>,
+        ghost_state: &mut Tracked<GhostState<'a, T>>,
+    )
+        requires
```

Comment:
Same thing, we should require that the list is well-formed. I think we need to do that for every function

Reply:
I mentioned on Slack that I encountered problem when trying to pass `old(ghost_state)` as an argument to `well_formed_list`. It kept complaining that ghost_state should be mutable reference but it indeed is as in the function parameter. Maybe it gets rid of `Tracked` automatically? I do not really know the correct way to do this. So I manually inline the `well_formed_list` spec function in the `requires`.

Reply:
If I understand correctly this error should not happen:

```
error: a mutable reference is expected here
   --> kernel/src/collections/list_v.rs:329:17
    |
329 |             old(self).well_formed_list(old(ghost_state)),
    |                 ^^^^

error: aborting due to 1 previous error
```

If it's the case you should either ask on the Verus Zillow or post an issue on github if you are convinced its a bug.
In the meanwhile add a comment with the name `VERUS-TODO` so that we know why we are doing that.